File: Code/2187/2187.cpp
/******************************************************************************
 * 文件名: 2187
 * 作者: BOLUN XU
 * 创建日期: 2025
 * 版本: 1.0
 * 描述: 完成旅途的最少时间 —— 二分写法。
 * 时间复杂度： O(Nlogn)
 * 空间复杂度： O(1)
 * 执行用时： 85 ms
 * 消耗内存： 96.10 mb
 ******************************************************************************/
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
  long long minimumTime(vector<int>& time, int totalTrips) {
    long long left = 0;
    long long right = (long long)*max_element(time.begin(), time.end()) * totalTrips + 1;
    // cout << "Debug: Initial bound: " << left << ", " << right << endl;

    while (left < right) {
      long long mid = left + (right - left) / 2;
      long long total = 0;

      for (int t:time) {
        total += mid/t;
      }
      // cout << "Debug: At time " << mid << ", total finish " << total << "trips" << endl;

      if (total >= totalTrips) {
        right = mid;
      }
      else {
        left = mid + 1;
      }
      // cout << "Debug: New bound " << left << ", " << right << endl;
    }
    return right;
  }
};

int main() {
  Solution s;
  vector<int> time{1,2,3};
  cout << s.minimumTime(time, 5) << endl;
}

File: Code/2187/2187.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
文件名: PersonalLeetCode
作者: Bolun Xu
创建日期: 2025
版本: 1.0
描述: 完成旅途的最少时间 —— 递归写法。
时间复杂度：
空间复杂度：
执行用时：
消耗内存：
"""
from typing import List


class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        def dfs(t, bustotal):
            # 计算当前时间 t 下每辆公交车的运行趟数
            for i in range(len(bustotal)):
                bustotal[i] = t // time[i]
            # print(f"Debug: At time {t}, the buses run {bustotal}")

            # 计算总趟数
            total = sum(bustotal)

            # 如果总趟数大于等于目标趟数，返回当前时间 t
            if total >= totalTrips:
                return t
            # 否则，继续递归，时间加 1
            else:
                return dfs(t + 1, bustotal)

        # 初始化 bustotal 数组
        bustotal = [0] * len(time)
        # 从时间 t = 1 开始递归
        return dfs(1, bustotal)


if __name__ == '__main__':
    sol = Solution()
    print(sol.minimumTime([1, 2, 3], 5))
    print(sol.minimumTime([5, 10, 10], 9))
